parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +628,349 @@ protected GroupType build(String name) {
           "Cannot build an empty group");
       return new GroupType(repetition, name, originalType, fields, id);
     }
+
+    public MapBuilder<Q> map(
+        Type.Repetition repetition) {
+      return new MapBuilder<Q>(self()).repetition(repetition);
+    }
+
+    public MapBuilder<Q> requiredMap() {
+      return new MapBuilder<Q>(self())
+          .repetition(Type.Repetition.REQUIRED);
+    }
+
+    public MapBuilder<Q> optionalMap() {
+      return new MapBuilder<Q>(self())
+          .repetition(Type.Repetition.OPTIONAL);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> list(Type.Repetition repetition) {
+      return new ListBuilder<BaseGroupBuilder<P, Q>>(this).repetition(repetition);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> requiredList() {
+      return list(Type.Repetition.REQUIRED);
+    }
+
+    public ListBuilder<BaseGroupBuilder<P, Q>> optionalList() {
+      return list(Type.Repetition.OPTIONAL);
+    }
+  }
+
+  /**
+   * A builder for {@link GroupType} objects.
+   *
+   * @param <P> The type that this builder will return from
+   *          {@link #named(String)} when the type is built.
+   */
+  public static class GroupBuilder<P> extends BaseGroupBuilder<P, GroupBuilder<P>> {
+
+    private GroupBuilder(P parent) {
+      super(parent);
+    }
+
+    private GroupBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected GroupBuilder<P> self() {
+      return this;
+    }
+  }
+
+  public static class MapBuilder<P> extends Builder<MapBuilder<P>, P> {
+    private static final Type STRING_KEY = Types
+        .required(PrimitiveTypeName.BINARY).as(OriginalType.UTF8).named("key");
+
+    public static class MapKeyBuilder<P> extends BasePrimitiveBuilder<P, MapKeyBuilder<P>> {
+      private MapBuilder<P> parent = null;
+
+      public MapKeyBuilder(MapBuilder<P> parent, PrimitiveTypeName type) {
+        super(parent.parent, type);
+        this.parent = parent;
+        repetition(Type.Repetition.REQUIRED);
+      }
+
+      public MapKeyBuilder(MapBuilder<P> parent, Class<P> returnType, PrimitiveTypeName type) {
```

Comment:
It is being passed to the parent. If we remove this then the following won't work.

 Types.requiredMap()
        .key(INT64)
        .requiredValue(INT64)
        .named("myMap");

I think returnType is redundant as it can be obtained from parent. However, if I remove returnType, the signature will conflict with the other constructor.